File: Max_Subarray_53.py
# Time - O(n)
# Space - O(n)
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        dp = [0 for i in range(len(nums)+1)]
#         Take a DP table and keep filling the table with the maximum sum of the subarray
        
        dp[0] = nums[0]
        res = nums[0]
        
        for i in range(1,len(nums)):
            dp[i] = max(dp[i-1] + nums[i], nums[i])
            res = max(res,dp[i])
        # print(dp)
        return res
    
    
        

        
        
# Time - O(n)
# Space - O(1)
#         sol_max = float('-inf')
#         curr_max =  float('-inf')
        
        
#         for i in range(len(nums)):
#             curr_max = max(nums[i] + curr_max, nums[i])
            
#             sol_max = max(sol_max, curr_max)
            
#         return sol_max

File: MaxSubArray53.java
//time complexity : O(n)
//space complexity : O(1)
//leet code : yes
//steps : maintain resting sum and total sum, run over itertations and pair up conseqtive indexs and sum them. and accordingly compare with global max. 

class MaxSubArray53{

        public int maxSubArray(int[] nums) {
            
            if(nums.length==0) return Integer.MIN_VALUE;
            int rSum = nums[0];
            int res = nums[0];
            for(int i=1; i<nums.length; i++){
               
                rSum = Math.max(rSum+nums[i],nums[i]);
                res = Math.max(rSum,res);
            }

            return res;
            
        }
    

    public static void main(String[] args) {
        
        MaxSubArray53 mx = new MaxSubArray53();
        System.out.println(mx.maxSubArray(new int[] {-2,1,-3,4,-1,2,1,-5,4}));
    }                                                             

}